File: app/src/main/java/com/iamwent/diary/modules/day/DayActivity.java
package com.iamwent.diary.modules.day;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import com.iamwent.diary.R;
import com.iamwent.diary.data.DiaryRepository;
import com.iamwent.diary.data.bean.Diary;
import com.iamwent.diary.modules.editor.EditorActivity;
import com.iamwent.diary.modules.preview.PreviewActivity;
import com.iamwent.diary.utils.LunarUtil;
import com.iamwent.diary.widget.OnRecyclerItemClickedListener;

import java.util.Collections;
import java.util.List;

import uk.co.chrisjenx.calligraphy.CalligraphyContextWrapper;

public class DayActivity extends Activity {
    private static final String EXTRA_YEAR = "year";
    private static final String EXTRA_MONTH = "month";

    private RecyclerView recyclerView;
    private List<Diary> diaries = Collections.emptyList();
    private DayListAdapter adapter;
    private int year;
    private int month;

    public static void start(Context context, int year, int month) {
        Intent starter = new Intent(context, DayActivity.class);
        starter.putExtra(EXTRA_YEAR, year);
        starter.putExtra(EXTRA_MONTH, month);
        context.startActivity(starter);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_day);

        year = getIntent().getIntExtra(EXTRA_YEAR, 0);
        month = getIntent().getIntExtra(EXTRA_MONTH, 0);

        TextView titleYear = (TextView) findViewById(R.id.tv_year);
        TextView titleMonth = (TextView) findViewById(R.id.tv_month);
        recyclerView = (RecyclerView) findViewById(R.id.day);

        recyclerView.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, true));
        recyclerView.addOnItemTouchListener(new OnRecyclerItemClickedListener(DayActivity.this) {
            @Override
            protected void onItemClick(View view, int position) {
                PreviewActivity.start(DayActivity.this, diaries.get(position).id);
            }
        });
        adapter = new DayListAdapter(this, diaries);
        recyclerView.setAdapter(adapter);

        titleYear.setText(String.format("%s年", LunarUtil.year2Chinese(year)));
        titleMonth.setText(String.format("%s月", LunarUtil.month2Chinese(month)));

        findViewById(R.id.btn_compose).setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        EditorActivity.start(DayActivity.this);
                    }
                });
    }

    @Override
    protected void onResume() {
        super.onResume();

        diaries = DiaryRepository.getInstance(this).queryDaysByYearAndMonth(year, month);
        adapter = new DayListAdapter(this, diaries);
        recyclerView.setAdapter(adapter);
    }

    @Override
    protected void attachBaseContext(Context newBase) {
        super.attachBaseContext(CalligraphyContextWrapper.wrap(newBase));
    }
}

File: app/src/main/java/com/iamwent/diary/modules/day/DayActivity.kt
package com.iamwent.diary.modules.day

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.activity.ComponentActivity
import androidx.activity.viewModels
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import com.iamwent.diary.R
import com.iamwent.diary.data.DiaryRepository
import com.iamwent.diary.databinding.ActivityDayBinding
import com.iamwent.diary.modules.DiaryViewModel
import com.iamwent.diary.modules.editor.EditorActivity
import com.iamwent.diary.modules.preview.PreviewActivity
import com.iamwent.diary.utils.LunarUtil
import com.iamwent.diary.widget.OnRecyclerItemClickedListener
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import javax.inject.Inject

@AndroidEntryPoint
class DayActivity : ComponentActivity() {

    private val binding: ActivityDayBinding by lazy {
        DataBindingUtil.setContentView(this, R.layout.activity_day)
    }
    private val adapter by lazy { DayListAdapter() }
    private val diaryViewModel by viewModels<DiaryViewModel>()

    private val year by lazy {
        intent.getIntExtra(EXTRA_YEAR, 0)
    }
    private val month by lazy {
        intent.getIntExtra(EXTRA_MONTH, 0)
    }

    @Inject
    lateinit var diaryRepository: DiaryRepository

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding.days.layoutManager =
            LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, true)
        binding.days.addOnItemTouchListener(object :
            OnRecyclerItemClickedListener(this@DayActivity) {
            override fun onItemClick(view: View?, position: Int) {
                adapter.currentList.getOrNull(position)?.let { diary ->
                    PreviewActivity.start(this@DayActivity, diary)
                }
            }
        })
        binding.days.adapter = adapter
        binding.year.text = "%s年".format(LunarUtil.year2Chinese(year))
        binding.month.text = "%s月".format(LunarUtil.month2Chinese(month))
        binding.compose.setOnClickListener {
            EditorActivity.start(this)
        }

        lifecycleScope.launch {
            diaryViewModel.getDiaries(year, month).collectLatest { diaries ->
                adapter.submitList(diaries)
            }
        }
    }

    companion object {
        private const val EXTRA_YEAR = "year"
        private const val EXTRA_MONTH = "month"
        fun start(context: Context, year: Int, month: Int) {
            val starter = Intent(context, DayActivity::class.java)
            starter.putExtra(EXTRA_YEAR, year)
            starter.putExtra(EXTRA_MONTH, month)
            context.startActivity(starter)
        }
    }
}
